File: LeetCode/992/main.go
package main

func atMost(arr []int, k int) int {
	n := len(arr)

	left, right, cnt, ans := 0, 0, 0, 0
	mp := make(map[int]int)

	for right < n {

		if mp[arr[right]] == 0 {
			cnt++
		}
		mp[arr[right]]++

		for cnt > k {
			mp[arr[left]]--
			if mp[arr[left]] == 0 {
				cnt--
			}
			left++
		}

		ans += right - left + 1
		right++

	}

	return ans
}

func subarraysWithKDistinct(nums []int, k int) int {
	return atMost(nums, k) - atMost(nums, k-1)
}

File: LeetCode/992/main.cpp
class Solution {
public:
  int atMost(vector<int> &arr, int k) {
    int n = arr.size();
    int left = 0, right = 0, cnt = 0, ans = INT_MIN;
    unordered_map<int, int> mp;

    while (right < n) {

      if (mp[arr[right]] == 0)
        cnt++;
      mp[arr[right]]++;

      while (cnt > k) {
        mp[arr[left]]--;
        if (mp[arr[left]] == 0)
          cnt--;
        left++;
      }

      ans += right - left + 1;
      right++;
    }

    return ans;
  }

  int subarraysWithKDistinct(vector<int> &arr, int k) {
    return atMost(arr, k) - atMost(arr, k - 1);
  }
};